sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvaluatePython.scala

```diff
@@ -120,10 +121,18 @@ object EvaluatePython {
     case (c: java.math.BigDecimal, dt: DecimalType) => Decimal(c, dt.precision, dt.scale)
 
     case (c: Int, DateType) => c
+    // Pyrolite will unpickle a Python datetime.date to a java.util.Calendar
+    case (c: Calendar, DateType) => DateTimeUtils.fromJavaCalendarForDate(c)
```

Comment:
How about we return `null` in this case? Other cases seems also returning `null` if it fails to be converted:

```
>>> from pyspark.sql.functions import udf
>>> f = udf(lambda x: x, "double")
>>> spark.range(1).select(f("id")).show()
+------------+
|<lambda>(id)|
+------------+
|        null|
+------------+
```

Seems we can do it like:

```scala

    case StringType => (obj: Any) => nullSafeConvert(obj) {
      case c: Calendar => null
      case _ => UTF8String.fromString(obj.toString)
    }
```

Reply:
Yea it's consistent with other un-convertible cases, but `StringType` is the default return type, I'm afraid many users many hit this and get confused.

Reply:
Right. Let's go ahead for 2. then. I am fine if it's done as an exception for practical purpose. Maybe we could add an if `isinstance(.., basestring)` and return directly as a shortcut. I haven't checked the perf diff but I think we can do it easily via profile as I mentioned above.

Reply:
WDYT about ^ @ueshin?

Reply:
Yeah, 2. should work for `StringType`.

I'd also like to add some documents like 1. for users to be careful about the return type. I've found that `udf`s return `null` and `pandas_udf`s throw some exception in most case when the return type is mismatching.
Of course we can try to make the behavior differences between `udf` and `pandas_udf` closer as possible in the future, but I think it is the best effort basis for the mismatching return type.

Reply:
My 2 cents:
* I am +0 for 2. I think having type coercion for str type but not for other types can be confusing to users. Realistically, I think any udf users would hit the `null` result for mismatch type anyway (I have hit it so many times..) and will learn that `null` means type mismatch. Even we make the behavior for str a bit friendly, they will likely to hit the issue with other types anyway. I don't think "default returnType for udf is str" is a strong reason for "having special type coercion for str", they seem orthogonal. 
**I prefer that we keep the type mismatch behavior consistent for str types vs other types (return null) and document this more clearly.**
* I find returning `null` for type mismatch is an unintuitive behavior of row-at-a-time udf and prefer not to replicate it in pandas udf.